File: 04-structural-patterns/03-composite/03-composite-qb.py
""" Convert the query builder to adapt fluent interface """

from abc import ABC, abstractmethod

class Expression(ABC):
    @abstractmethod
    def to_sql(self):
        pass

class Condition(Expression):
    def __init__(self, field, operator, value):
        self.field = field
        self.operator = operator
        self.value = value

    def to_sql(self):
        return f"{self.field} {self.operator} '{self.value}'"

class And(Expression):
    def __init__(self, *children):
        self.children = children

    def to_sql(self):
        return "(" + " AND ".join(child.to_sql() for child in self.children) + ")"

class Or(Expression):
    def __init__(self, *children):
        self.children = children

    def to_sql(self):
        return "(" + " OR ".join(child.to_sql() for child in self.children) + ")"

class QueryBuilder:
    def __init__(self, table, columns=None):
        self.table = table
        self.columns = columns or ["*"]
        self.condition = None

    def where(self, expr):
        self.condition = expr
        return self

    def build(self):
        sql = f"SELECT {', '.join(self.columns)} FROM {self.table}"
        if self.condition:
            sql += f" WHERE {self.condition.to_sql()}"
        return sql + ";"

if __name__ == "__main__":
    cond1 = Condition("age", ">", "30")
    cond2 = Condition("city", "=", "Delhi")
    cond3 = Condition("salary", ">", "50000")

    expression = Or(
        And(cond1, cond2),
        cond3
    )

    qb = QueryBuilder("employees", ["id", "name", "salary"]).where(expression)
    print(qb.build())

File: 04-structural-patterns/03-composite/03-composite-qb.go
package main

import (
	"fmt"
	"strings"
)

// Expression represents part of a WHERE clause.
type Expression interface {
	ToSQL() string
}

type Condition struct {
	Field, Operator, Value string
}

func (c *Condition) ToSQL() string {
	return fmt.Sprintf("%s %s '%s'", c.Field, c.Operator, c.Value)
}

type And struct {
	Expressions []Expression
}

func (a *And) ToSQL() string {
	parts := make([]string, len(a.Expressions))
	for i, expr := range a.Expressions {
		parts[i] = expr.ToSQL()
	}
	return "(" + strings.Join(parts, " AND ") + ")"
}

type Or struct {
	Expressions []Expression
}

func (o *Or) ToSQL() string {
	parts := make([]string, len(o.Expressions))
	for i, expr := range o.Expressions {
		parts[i] = expr.ToSQL()
	}
	return "(" + strings.Join(parts, " OR ") + ")"
}

// QueryBuilder builds SELECT queries using the expression tree.
type QueryBuilder struct {
	Table     string
	Columns   []string
	Condition Expression
}

func (qb *QueryBuilder) Build() string {
	cols := "*"
	if len(qb.Columns) > 0 {
		cols = strings.Join(qb.Columns, ", ")
	}
	sql := fmt.Sprintf("SELECT %s FROM %s", cols, qb.Table)
	if qb.Condition != nil {
		sql += " WHERE " + qb.Condition.ToSQL()
	}
	return sql + ";"
}

func main() {
	// Example
	cond1 := &Condition{"age", ">", "30"}
	cond2 := &Condition{"city", "=", "Delhi"}
	cond3 := &Condition{"salary", ">", "50000"}

	expr := &Or{
		Expressions: []Expression{
			&And{[]Expression{cond1, cond2}},
			cond3,
		},
	}

	qb := &QueryBuilder{
		Table:     "employees",
		Columns:   []string{"id", "name", "salary"},
		Condition: expr,
	}

	fmt.Println(qb.Build())
}
